### DISCUSSION TITLE
Multipackage Bundles: MERGE doesn't reduce the size of the executables

### DISCUSSION THREAD

**Discussion by mush42** [pyinstaller/pyinstaller#6402] [PyInstaller] [answered]
Hi,

I'm using PyInstaller to bundle blindpandas/bookworm on Windows with Python 3.9.

The main script that launches the program and the one that pyinstaller builds depend  on the core code packaged as python **wheel** which is installed to the site-packages directory before running pyinstaller.

Recently I started to write small utility scripts that perform small tasks and exit, those scripts depend on the same package that the main script depends on. They share 100% of the code with the main package with no extras.

Actually, all of the scripts are thin wrappers around **bookworm** package.

Using the Multipackage Bundle feature of PyInstaller via MERGE to build the executables yields large executables as follows:
* Bookworm.exe: 8531 KB
* document_indexer.exe: 7227 KB

I invoke pyinstaller using this command:

```bash
pyinstaller Bookworm.spec --clean -y --distpath Bookworm
```

Are the given executable sizes expected? or should I get significantly smaller executables for utility scripts?
 
Here are the scripts:

**launcher.py** (the main program)

```python
# coding: utf-8

if __name__ == "__main__":
    import multiprocessing
    multiprocessing.freeze_support()
    from bookworm import bookworm
    bookworm.main()
```

**document_indexer.py** (A small utility script to add the given filename to the full-text-search database)
```python
# coding: utf-8


if __name__ == "__main__":
    import multiprocessing
    multiprocessing.freeze_support()
    from bookworm.bookshelf import commandline_app
    commandline_app.main()
```

And here is the **spec** file I use to build the executables:

```python
# -*- mode: python ; coding: utf-8 -*-

from pathlib import Path
from PyInstaller.utils.hooks import collect_data_files, collect_submodules


# Data files
PACKAGES_WITH_DATA = [
    "justext",
    "tld",
    "docx",
    "pptx",
]
BOOKWORM_RESOURCES = collect_data_files(
    "bookworm",
    excludes=[
        "*.po",
        "*.md",
    ],
)
DATA_FILES = [
    (
        src,
        Path(dst).relative_to("bookworm"),
    )
    for src, dst in BOOKWORM_RESOURCES
]
for pkg_name in PACKAGES_WITH_DATA:
    DATA_FILES += collect_data_files(pkg_name)

# Hidden imports
HIDDEN_SUBMODULES = [
    "babel",
    "odf",
]
HIDDEN_IMPORTS = [
    "numpy",
    "cv2",
    "pkg_resources.py2_warn",
]
for package_with_submodules in HIDDEN_SUBMODULES:
    HIDDEN_IMPORTS += collect_submodules(package_with_submodules)

block_cipher = None

bookworm_a = Analysis(
    ["launcher.py"],
    pathex=[""],
    datas=DATA_FILES,
    binaries=[],
    # See: https://stackoverflow.com/questions/37815371/
    hiddenimports=HIDDEN_IMPORTS,
    hookspath=[],
    runtime_hooks=[],
    excludes=[
        "tkinter",
    ],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)

document_indexer_a = Analysis(
    ["document_indexer.py"],
    pathex=[],
    binaries=[],
    datas=[],
    hiddenimports=[],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)

MERGE(
    (bookworm_a, 'launcher', 'Bookworm'),
    (document_indexer_a, 'document_indexer', 'document_indexer'),
)

bookworm_pyz = PYZ(bookworm_a.pure, bookworm_a.zipped_data, cipher=block_cipher)
bookworm_exe = EXE(
    bookworm_pyz,
    bookworm_a.scripts,
    bookworm_a.dependencies,
    exclude_binaries=True,
    name="Bookworm",
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=False,
    console=False,
    version="assets\\version_info.txt",
    icon="assets\\bookworm.ico",
)

document_indexer_pyz = PYZ(
    document_indexer_a.pure, document_indexer_a.zipped_data, cipher=block_cipher
)

document_indexer_exe = EXE(
    document_indexer_pyz,
    document_indexer_a.scripts,
    document_indexer_a.dependencies,
    exclude_binaries=True,
    name="document_indexer",
    debug=False,
    bootloader_ignore_signals=False,
    strip=True,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
    icon="assets\\bookworm.ico",
)

col = COLLECT(
    bookworm_exe,
    bookworm_a.binaries,
    bookworm_a.zipfiles,
    bookworm_a.datas,
    document_indexer_exe,
    document_indexer_a.binaries,
    document_indexer_a.zipfiles,
    document_indexer_a.datas,
    strip=False,
    upx=False,
    upx_exclude=[],
    name="Bookworm",
)
```

**Comment by rokm** [comment, marked as answer]
> or should I get significantly smaller executables for utility scripts?

In the presented scenario, I think the answer is no. In fact, using `MERGE` even worsens your situation...

You are freezing your two scripts in `onedir` mode, where the expected executable size is the size of the bootloader + size of the PYZ archive (where byte-compiled .pyc files are collected), because all other binaries and data files are external and are collected into the application's directory.

`MERGE` does not affect PYZ archive, only the non-PYZ resources (binaries and data). So for example, if you had two `onefile` programs, you could use `MERGE` to reduce size of one of them by moving all non-PYZ files into the other program and referencing them. Or, if you had two `onedir` programs that are collected into separate (!) dist directories, one of them could have its overall size reduced by moving all external files into the other directory and referencing them. Or, you could have a mixture of `onedir` and `onefile` programs where binaries and data files are collected in one of the programs, and other programs reference that. But this comes with a price, because `onedir` programs that have dependency forward references due to `MERGE` now behave like `onefile` programs, in the sense that they set up temporary directory and extract/collect all required external files into it (depending on combination, this involves copying data from primary onedir program's directory, or extracting it from primary onefile program, but in general, it slows down the program's startup). Also, the forward references need to be written to `onedir` program's TOC, so the `onedir` executable is in fact slightly larger after the MERGE (but that should be noticeable only with applications with large amount of collected data files and binaries).

Now, back to your scenario. You have two `onedir` programs, and you collect them into same dist directory (using joint `COLLECT`). This on its own already ensures that everything that can be shared (collected data files and binaries) is shared and collected only in one copy. The executables themselves should be of comparable size to those of individual `onedir` builds, i.e. the bootloader + PYZ archive (you can find the PYZ archive file in the `build` directory, if you want to check its size); because MERGE does not affect PYZ, you gain nothing by using it. Even worse, one of your executables now needs to collect all external data files and binaries into temporary directory before it can start.

So at the end of the day, a joint COLLECT without MERGE is probably the best option here. Neither technique (joint COLLECT or MERGE) does anything about PYZ archive and individual onedir executable size (i.e., sharing of collected pure-python modules) - I think if you wanted to optimize that, you would need to build with `noarchive` option in `Analysis` set to `True`, so that collected `.pyc` modules are collected as external files instead of executable-specific PYZ archive (so that a single copy is shared among both executables as well).

**Comment by mush42** [comment]
Thanks @rokm  for the clarification. 

Is there any way to have pure python dependencies collected to a single archive (external file) that is then shared by the executables?

**Reply by rokm** [reply]
No, there is not. The only way to achieve sharing of pure-python dependencies right now is to use `noarchive=True` so that .pyc modules are not collected into PYZ archive(s).

On the other hand, do you need separate executables for your program(s)? Your original example could as well be a single executable that executes either the launcher codepath or the document_indexer codepath based on the first argument it is passed (e.g., `single-program launcher <the-rest-of-arguments` vs `single-program document-indexer <the-rest-of-arguments>`). And then you can use simple .bat files with original names (`bookworm.bat`, `document_indexer.bat`) that invoke the unified executable in the corresponding mode.

**Reply by mush42** [reply]
That is a good option. But I warry about startup time, it is already slow as is. 
Doesn't a single file executable take longer to start?

**Reply by rokm** [reply]
I meant a single `onedir` executable/program, not a `onefile` one.

**Reply by mush42** [reply]
That was very helpful. Thanks!
